src/android/l2cap_channel.rs

```diff
@@ -0,0 +1,245 @@
+        });
+
+        thread::spawn(move || {
+            debug!("l2cap write thread running!");
+
+            output_stream.vm().with_env(|env| {
+                let stream = output_stream.as_local(env);
+
+                loop {
+                    match write_receiver.recv_blocking() {
+                        Err(e) => {
+                            warn!("failed to dequeue l2cap packet to send: {:?}", e);
+                            break;
+                        }
+                        Ok(packet) => {
+                            let b = PrimitiveArray::from(env, u8toi8(&packet));
+                            if let Err(e) = stream.write_byte_array(Some(&*b)) {
+                                warn!("failed to write to l2cap channel: {:?}", e);
+                                break;
+                            };
+                        }
+                    }
+                }
+            });
+
+            debug!("l2cap write thread exiting!");
+        });
+
+        Ok((
+            L2capChannelReader {
+                closer: closer.clone(),
+                stream: read_receiver,
+            },
+            L2capChannelWriter {
+                closer,
+                stream: write_sender,
+            },
+        ))
+    })
+}
+
+/// Utility struct to close the channel on drop.
+pub(super) struct L2capCloser {
+    channel: Global<BluetoothSocket>,
+}
+
+impl L2capCloser {
+    fn close(&self) {
+        self.channel.vm().with_env(|env| {
+            let channel = self.channel.as_local(env);
+            match channel.close() {
+                Ok(()) => debug!("l2cap channel closed"),
+                Err(e) => warn!("failed to close channel: {:?}", e),
+            };
+        });
+    }
+}
+
+impl Drop for L2capCloser {
+    fn drop(&mut self) {
+        self.close()
+    }
+}
+
+pub struct L2capChannelReader {
+    stream: Receiver<Vec<u8>>,
+    closer: Arc<L2capCloser>,
+}
+
+impl L2capChannelReader {
+    #[inline]
+    pub async fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        let packet = self
+            .stream
+            .recv()
+            .await
+            .map_err(|_| Error::new(ErrorKind::ConnectionFailed, None, "channel is closed".to_string()))?;
+
+        if packet.len() > buf.len() {
+            return Err(Error::new(
+                ErrorKind::InvalidParameter,
+                None,
+                "Buffer is too small".to_string(),
+            ));
+        }
+
+        buf[..packet.len()].copy_from_slice(&packet);
+
+        Ok(packet.len())
+    }
+
+    #[inline]
+    pub fn try_read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        let packet = self.stream.try_recv().map_err(|e| match e {
+            TryRecvError::Empty => Error::new(ErrorKind::NotReady, None, "no received packet in queue".to_string()),
+            TryRecvError::Closed => Error::new(ErrorKind::ConnectionFailed, None, "channel is closed".to_string()),
+        })?;
```

Comment:
What would you think about also returning `e` as the source for these errors (e.g. `Error::new(..., Some(Box::new(e)), ...)`)?

Reply:
Makes sense, I just wanted to ask the question. Because the source is in a `Box<dyn Error>` it doesn't really become part of the public API, IMO, but it doesn't provide much benefit either.